flink-tests/src/test/java/org/apache/flink/api/datastream/DataStreamBatchExecutionITCase.java

```diff
@@ -0,0 +1,159 @@
+import org.apache.flink.api.common.time.Time;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.configuration.ExecutionOptions;
+import org.apache.flink.runtime.testutils.MiniClusterResourceConfiguration;
+import org.apache.flink.streaming.api.datastream.DataStreamSource;
+import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+import org.apache.flink.test.util.MiniClusterWithClientResource;
+
+import org.junit.ClassRule;
+import org.junit.Ignore;
+import org.junit.Test;
+
+import java.util.List;
+
+import static org.hamcrest.Matchers.contains;
+import static org.junit.Assert.assertThat;
+
+/**
+ * Integration test for {@link org.apache.flink.api.common.RuntimeExecutionMode#BATCH} execution on
+ * the DataStream API.
+ *
+ * <p>We use a {@link MiniClusterWithClientResource} with a single TaskManager with 1 slot to verify
+ * that programs in BATCH execution mode can be executed in stages.
+ */
+public class DataStreamBatchExecutionITCase {
+	private static final int DEFAULT_PARALLELISM = 1;
+
+	@ClassRule
+	public static MiniClusterWithClientResource miniClusterResource = new MiniClusterWithClientResource(
+			new MiniClusterResourceConfiguration.Builder()
+					.setNumberTaskManagers(1)
+					.setNumberSlotsPerTaskManager(DEFAULT_PARALLELISM)
+					.build());
+
+	/**
+	 * We induce a failure in the last mapper. In BATCH execution mode the part of the pipeline
+	 * before the key-by should not be re-executed. Only the part after that will restart. We check
+	 * that by suffixing the attempt number to records and asserting the correct number.
+	 */
+	@Test
+	public void batchFailoverWithKeyByBarrier() throws Exception {
+
+		final StreamExecutionEnvironment env = getExecutionEnvironment();
+
+		DataStreamSource<String> source = env.fromElements("ciao", "ciao");
+
+		SingleOutputStreamOperator<String> mapped = source
+				.map(new SuffixAttemptId("a"))
+				.map(new SuffixAttemptId("b"))
+				.keyBy(in -> in)
+				.map(new SuffixAttemptId("c"))
+				.map(new OnceFailingMapper("d"));
+
+		List<String> result = mapped.executeAndCollect(2);
+
+		// only the operators after the key-by "barrier" are restarted and will have the "attempt 1"
+		// suffix
+		assertThat(result, contains("ciao-a0-b0-c1-d1", "ciao-a0-b0-c1-d1"));
+	}
+
+	@Ignore
+	@Test
+	public void stagedExecutionWithDifferentSlotGroups() throws Exception {
+
+		final StreamExecutionEnvironment env = getExecutionEnvironment();
+
+		DataStreamSource<String> source = env.fromElements("ciao", "ciao");
+
+		SingleOutputStreamOperator<String> mapped = source
+				.map(new SuffixAttemptId("a"))
+				.map(new SuffixAttemptId("b"))
+				.map(new SuffixAttemptId("c"))
+				.slotSharingGroup("foo")
+				.map(new SuffixAttemptId("d"))
+				.slotSharingGroup("bar");
+
+		List<String> result = mapped.executeAndCollect(2);
```

Comment:
nit: Do we need to add the limit? I think it's an unnecessary trap, if we add more incoming records in the test.

Reply:
If I don't use a limit I have to use use the version that returns an `Iterator` and I manually have to deal with that. Though with `assertThat()` it should be pretty smooth. Let me try it.

Reply:
Ah, ok I have missed the different return type. I am fine with using the limit then.

Another option is to use `CollectionUtil.iteratorToList()`

Reply:
I'm doing that now, thanks!